src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -171,6 +171,247 @@ cudnnSetOpTensorDescriptor(opTensorDesc,
+                CoordinateDiff padding_above = convolution->get_padding_above_forward();
+
+                if (padding.size() > 3)
+                {
+                    throw std::runtime_error(node->get_name() +
+                                             "with more than 3D is not implemented.");
+                }
+                for (auto a : data_dilation_strides)
+                {
+                    if (a != 1)
+                    {
+                        throw std::runtime_error(node->get_name() +
+                                                 "with data dilation is not implemented.");
+                    }
+                }
+                for (int i = 0; i < padding.size(); i++)
+                {
+                    if (padding[i] != padding_above[i])
+                    {
+                        throw std::runtime_error(node->get_name() +
+                                                 "with asymmetric padding is not implemented.");
+                    }
+                }
+
+                writer.block_begin("  // " + node->get_name());
+                writer << "float alpha = 1.0;\n";
+                writer << "float beta = 0.0;\n";
+
+                // construct input and output tensor descriptor
+                kernel::emit_cudnnFilterDescriptor(
+                    writer, args0, tensor_format, data_type, args[0].get_shape());
+                kernel::emit_cudnnTensorDescriptor(
+                    writer, args1, tensor_format, data_type, args[1].get_shape());
+                kernel::emit_cudnnTensorDescriptor(
+                    writer, out0, tensor_format, data_type, out[0].get_shape());
+                kernel::emit_cudnnConvolutionDescriptor(writer,
+                                                        conv_descriptor,
+                                                        padding,
+                                                        window_movement_strides,
+                                                        window_dilation_strides,
+                                                        mode,
+                                                        data_type);
+
+                writer << "void* workspace = "
+                          "runtime::gpu::create_gpu_buffer("
+                       << workSapceSizeInBytes << ");\n";
+
+                writer << "cudnnConvolutionBackwardData(cudnn_handle, "
+                       << "&alpha, " << args0 << ", " << args[0].get_name() << ", " << args1 << ", "
+                       << args[1].get_name() << ", " << conv_descriptor << ", " << conv_algo << ", "
+                       << "workspace, " << workSapceSizeInBytes << ", "
+                       << "&beta, " << out0 << ", " << out[0].get_name() << ");\n";
+
+                writer << "runtime::gpu::free_gpu_buffer(workspace);\n";
+                writer.block_end();
+            }
+
+            template <>
```

Comment:
I noticed that these three ops (Conv,BPData,BPFilters) all have essentially the same boiler-plate/prologue code for emitting them. This could be refactored into a single function for maintainability, but is maybe not a high priority for now.

Reply:
Right, most of the code are very similar. there are slight different with each other.
Since those are not completed yet(missing asymmetric padding and data dilation), let's do it later.